kloud: support importing credentials from a file

rjeczalik opened the pull request:
- adds `credential.*`, `import` and `machine.list` methods to kloud
- adds experimental commands `kd build -f`, `kd cred list`, `kd cred add` and `kd cred use`
## Description

The purpose of this PR is to add new API to the kloud to support importing stacks from a template file.

Last missing piece after merging https://github.com/koding/koding/pull/9052 and open-sourcing https://github.com/koding/kloud-provider-example is to allow for integration testing stack builds, in order to allow assessing contribution with new kloud providers.

The idea for integration tests is to extend kd with build subcommand, that would allow for directly building stack templates from command line.

The `kd build` is going to be used for integration tests to test building stack in non-interactive mode. 
## Motivation and Context

Motivation is to to allow for testing kloud providers. And indirectly support stack workflow in KD.
Context is continuation of https://github.com/koding/koding/pull/9052 and https://github.com/koding/koding/pull/9139 and a prerequisite for  [kloud provider issues](https://github.com/koding/koding/issues?q=is%3Aissue+is%3Aopen+support+label%3A%22Category-Stack+Support%22).

Related: https://github.com/koding/koding/issues/9024 (this PR partially implements autogenrated schema API)

Deps: https://github.com/koding/koding/pull/9166
## How Has This Been Tested?

Includes tests plus manual QA.
## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
## Checklist:
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

rjeczalik committed:
config: remove unneeded creds

rjeczalik committed:
kloud/stack: create API definition for KD backend

rjeczalik committed:
stack: implement CredentialDescribe kite method

rjeczalik committed:
credential: add Client API for managing credentials

rjeczalik committed:
credential: implement Validate backed by MongoDB

rjeczalik committed:
credential: implement Creds backed by MongoDB

rjeczalik committed:
credential: add Client implementation

rjeczalik committed:
credential: add debug logging

rjeczalik committed:
credential: implement SetCred backed by Mongo

rjeczalik committed:
stack: add CredentialList method

rjeczalik committed:
stack: add CredentialAdd method

rjeczalik committed:
stack: verify and bootstrap credential on CredentialAdd

rjeczalik committed:
klientctl: initial commit for credential import command

rjeczalik committed:
stack: fix generating schema description

rjeczalik committed:
klientctl/config: read configuration from external source

This PR makes KD / klient configurable, as oppose to
only having static configuration.

Currently kd and klient has configuration injected
from koding/config/config.json and during linking.

The configuration is built-in and cannot be changed.

This PR makes it possible to overwrite configuration
variables like KontrolURL or TunnelURL by creating
external source of configuration, a BoltDB database
located at ~/.config/koding/config.bolt.

KD / klient upon start still uses built-in configuration,
but additionally it overwrites built-in configuration
with the one loaded from BoltDB database.

The default directory where kd / klient looks for
config.bolt can be overwritten with KODING_HOME
environmental variable.

A database file is used instead of simple configuration
file since kd and klient share the same file,
and exclusive access to it ensure by use of BoltDB
file locking.

For now if the configuration database does not
exist it is created from builtin configuration.

Prerequisite for #8965, allows for removing all
other sources for configuration.

rjeczalik committed:
object: add Merge func

rjeczalik committed:
klientctl: add config subcommand

rjeczalik committed:
object: add (*Builder).Set for setting arbitrary fields

rjeczalik committed:
klientctl: fix unbound KD_DEBUG variable

rjeczalik committed:
kloud: fix for apply and plan methods

rjeczalik committed:
klientctl: add "credential import" subcommand

rjeczalik committed:
klientctl: add "credential list" subcommand

rjeczalik committed:
klientctl: add "credential create" subcommand

rjeczalik committed:
stack: do not validate empty bootstrap

rjeczalik committed:
stack: fix decoding credential payload

rjeczalik commented:
@cihangir @ppknap Please take a look. I'm going to send PR with`kloud.import` and `kd stack build` implementations separately, since this one is large enough.

rjeczalik committed:
provider/aws: add enums for Cred.Region

cihangir commented on go/src/koding/kites/kloud/stack/provider/apply.go:
```diff
@@ -36,20 +37,22 @@ import (
-	var arg stack.ApplyRequest
-	if err := bs.Req.Args.One().Unmarshal(&arg); err != nil {
-		return nil, err
+	arg, ok := ctx.Value(stack.ApplyRequestKey).(*stack.ApplyRequest)
```
why do you need this? i could find the usage of it in this PR

rjeczalik replied:
In this PR I call `HandleBootstrap` method. The `HandleBootstrap` method before the change read BootstrapRequest value from `(*kite.Request).Args`, after the change it looks it up in the `context.Context` first and use it instead if found. The change is required to make it possible to call `HandleBoostrap` outside kite context.

Also I've applied the same change to all `Handle*` methods for the sake of consistency.

cihangir commented on go/src/koding/kites/kloud/stack/provider/authenticate.go:
```diff
@@ -8,9 +8,13 @@ import (
-	var arg stack.AuthenticateRequest
-	if err := bs.Req.Args.One().Unmarshal(&arg); err != nil {
-		return nil, err
+	arg, ok := ctx.Value(stack.AuthenticateRequestKey).(*stack.AuthenticateRequest)
```
same here with prev comment

cihangir commented on go/src/koding/kites/kloud/stack/provider/builder.go:
```diff
@@ -398,43 +334,63 @@ func (b *Builder) MachineUIDs() map[string]string {
+	groupIDs := []bson.ObjectId{account.Id}
+
+	if groupname != "" {
+		// fetch jGroup from group slug name
```
lets keep function bodies in manageable length

rjeczalik replied:
This is a leftover from previous implementation, this method is going to be replaced with credential.Client introduced with this PR once it's feature-complete.

https://github.com/koding/koding/blob/990b6eff97f0de6a867a42147d2af0ce44df36de/go/src/koding/kites/kloud/stack/provider/builder.go#L338

rjeczalik replied:
I can also share my long term plans - previous implementation was just a few methods with hundreds of lines:

- https://github.com/koding/koding/commit/6c4720f95a7c3938e0fe3e3d427ca07ad50de6ad#diff-d3b87760e7a4403c45e7dc6e12e714c9L460
- https://github.com/koding/koding/commit/6c4720f95a7c3938e0fe3e3d427ca07ad50de6ad#diff-d3b87760e7a4403c45e7dc6e12e714c9L498
- etc.

I've introduced a Builder type just a temporary abstraction [in this PR](https://github.com/koding/koding/commit/6c4720f95a7c3938e0fe3e3d427ca07ad50de6ad). The purpose of the Builder was to create a base abstraction for validating user, validating credentials, building models, expanding user templates and updating stack models. Basically a facade, that used previous implementation (the global funcs), but stored the product of the methods within itself, so was slightly possible to mock it. The Builder was a first step to fix previous abstraction leak (or lack of abstraction). Right now I'm in an ongoing effort to refactor Builder into separate packages that are responsible for each step of the stack building process: credential, stack and terraform. Once it's done, the Builder will be gone.

The aim is to have functional tests for kloud / stack / template, in addition to integration ones with `kd`.

cihangir commented on go/src/koding/kites/kloud/stack/provider/describe.go:
```diff
@@ -0,0 +1,90 @@
+
+var describeTags = []string{"bson", "hcl", "json"}
+
+func Describe(v interface{}) ([]stack.Value, error) {
```
could you please write what does Describe do?

rjeczalik replied:
Done.

cihangir commented on go/src/koding/kites/kloud/stack/provider/describe_test.go:
```diff
@@ -0,0 +1,50 @@
+	}{
+		"AWS credential": {
+			&aws.Cred{},
+			[]stack.Value{{
```
could you test other types as well - not just strings

rjeczalik replied:
Done.

cihangir commented on go/src/koding/kites/kloud/stack/provider/provider.go:
```diff
@@ -51,6 +57,44 @@ func All() []*Provider {
+		}
+	}
+
+	n := len(nonempty)
```
you can replace lines between 76:81 with `desc := make(map[string]*stack.Description)`

rjeczalik replied:
Done.

cihangir commented on go/src/koding/kites/kloud/utils/object/merge.go:
```diff
@@ -0,0 +1,26 @@
+		return
+	}
+	vv1 := reflect.ValueOf(v1).Elem()
+	for i := 0; i < vv1.NumField(); i++ {
```
seems like this wont work with nested-structs

rjeczalik replied:
It does not set fields recursively, as it would make the implementation more complex and it's not needed for now.

Added a TODO.

cihangir commented on go/src/koding/kites/kloud/utils/object/object.go:
```diff
@@ -69,6 +72,66 @@ func (b *Builder) Build(v interface{}, ignored ...string) Object {
+// TODO(rjeczalik): add support for patching - accept path instead of key,
+// so it's possible to partially update nested structs / maps.
+func (b *Builder) Set(v interface{}, key string, value interface{}) error {
+	type setter interface {
```
👍

cihangir commented on go/src/koding/kites/kloud/utils/object/object.go:
```diff
@@ -69,6 +72,66 @@ func (b *Builder) Build(v interface{}, ignored ...string) Object {
 	return obj
 }
 
+// TODO(rjeczalik): add support for patching - accept path instead of key,
```
could you please write tests for this function with diff types?

rjeczalik replied:
If you mean https://github.com/koding/koding/pull/9182#discussion_r82774059 then it does not work for different types.

cihangir replied:
nope, i meant, integer, int64, float etc

cihangir commented on go/src/koding/klientctl/main.go:
```diff
@@ -49,6 +49,11 @@ var sudoRequiredFor = []string{
 // cli rewrite.
 var log logging.Logger
```
could you create this logger with debug parameter is if `KD_DEBUG` exists? and remove https://github.com/koding/koding/pull/9182/files#diff-9d1c92b05bf71fa597eb81b434721addR62
```
+	if debug {
 +		log.SetLevel(logging.DEBUG)
 +		k.SetLogLevel(kite.DEBUG)
 +	}
```

rjeczalik committed:
koding/config: move to koding/kites/config

rjeczalik committed:
modelhelper: return error directly

rjeczalik committed:
modelhelper: fix build

rjeczalik commented on go/src/koding/config/config.json:
```diff
@@ -55,6 +55,50 @@
         "url": "https://dev-p2.koding.com/-/ipcheck"
       }
     ],
+    "kd-latest": [
```
Pawel is already working on this (https://github.com/koding/koding/issues/8920).

ppknap replied:
Actually this part is done, I will move all these variables to main.{}.coffee. Haven't done this yet because I'm not sure if there is main.prod.coffee in credential(?) repo which will also require updates in that case.

rjeczalik commented on go/src/koding/config/cache.go:
```diff
@@ -0,0 +1,78 @@
+package config
```
Moved to koding/kites/config.

rjeczalik commented on go/src/koding/db/models/error.go:
```diff
@@ -0,0 +1,65 @@
+//
+// TODO(rjeczalik): the db/mongo helpers should be decorating errors
+// instead, so database handling layer is transparent and no
+// library specific errors (e.g. mgo.ErrNotFound, pq.Error) are leaked.
```
I've removed the comment - it was about making modelhelper package, not models one, do not return mgo.* types and values.

The idea behind this one was to decouple mongo / postgres from packages that require database access. For new package I try to achieve that by creating interfaces for database access, like:

- [credential.Store](https://github.com/koding/koding/blob/55dbb189af66bb1d5b6410bfc8568721e826b79b/go/src/koding/kites/kloud/credential/store.go#L67-L71)
- [credential.Database](https://github.com/koding/koding/blob/55dbb189af66bb1d5b6410bfc8568721e826b79b/go/src/koding/kites/kloud/credential/client.go#L60-L103)
- [provider.Database](https://github.com/koding/koding/blob/55dbb189af66bb1d5b6410bfc8568721e826b79b/go/src/koding/kites/kloud/stack/provider/database.go#L83-L108)

In the long run my aim is to make it possible to have tests for stack template building which do not require actual database and fixtures.

rjeczalik commented on go/src/koding/db/models/error.go:
```diff
@@ -0,0 +1,65 @@
+
+// ResError is a helper function that decorates the given err with more
+// meaningful error value, giving the caller context to recover.
+func ResError(err error, resource string) error {
```
I'm not sure what should change here.

The `ResError` is something like `NewNotFoundErrorIf`, it wraps the err only under given condition. I think \*Error types do not require any special treatment to make sure they are not misused (unexporting fields etc.) - if wrong value is created (e.g. `var err error = &NotFoundError{Err: nil}`) it's treated as programmer error (I tried to follow the assumption stdlib makes for it's own error types - e.g. [assumption that for any OpError values Err field is not-nil](https://github.com/golang/go/blob/dd307da10c3582c195928d9bf073d1b0b01f2135/src/net/net.go#L448)).

cihangir replied:
>  (unexporting fields etc.) 

this is my concern and it is not directed only for errors, if someone can misuse it they will.

rjeczalik commented on go/src/koding/db/mongodb/modelhelper/credentials.go:
```diff
@@ -25,6 +25,28 @@ func GetCredentialsFromIdentifiers(identifier ...string) ([]*models.Credential,
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("credentials lookup error: %v", err)
```
Done.

rjeczalik commented on go/src/koding/db/mongodb/modelhelper/credentials.go:
```diff
@@ -61,6 +83,12 @@ func UpdateCredential(identifier string, data bson.M) error {
 
+func CreateCredential(cred *models.Credential) error {
+	return Mongo.Run(CredentialsColl, func(c *mgo.Collection) error {
+		return c.Insert(cred)
```
Currently no Create\* function does that, they assume the caller is responsible for creating / passing valid models.\* struct. Could we leave that as-is to be consistent?

rjeczalik commented on go/src/koding/kites/kloud/credential/mongo.go:
```diff
@@ -106,3 +108,470 @@ func (db *mongoStore) Put(_ string, creds map[string]interface{}) error {
+	})
+}
+
+func match(existing, expected []string) bool {
```
The `match` does not test for equality here, so we can't `len(existing) != len(expected)` here.

I've changed the func name to `contains`.

rjeczalik committed:
credential: improve Perm interface, add inline docs

rjeczalik commented on go/src/koding/kites/kloud/stack/credential.go:
```diff
@@ -0,0 +1,249 @@
+}
+
+type Enumer interface {
+	Enum() []*Enum
```
Done.

rjeczalik committed:
stack: add inline docs for Credential* methods

rjeczalik commented on go/src/koding/kites/kloud/credential/mongo.go:
```diff
@@ -106,3 +108,470 @@ func (db *mongoStore) Put(_ string, creds map[string]interface{}) error {
+	return nil
+}
+
+func (db *mongoDatabase) Lock(cred *Cred) error {
```
`create`, `edit` and `remove` operations on a single credential are exclusive ones - while one is in progress, any other should be rejected.

I coded for future, since right now I need only create method for `kd stack create`. I've removed the locking and will re-introduce once other methods are needed.

rjeczalik committed:
stack: add more tests for Describe method

ppknap commented on go/src/koding/klientctl/config.go:
```diff
@@ -54,3 +58,82 @@ func createFolderAtHome(cf ...string) (string, error) {
+	}
+
+	if err := updateKonfigCache(c.Args().Get(0), c.Args().Get(1)); err != nil {
+		fmt.Println(err)
```
log.Error or maybe nothing since the error is returned to the caller

ppknap replied:
Oh I see, so maybe is it better to log to stderr like you do [here](https://github.com/koding/koding/pull/9182/files#diff-47e4ab0f8167fbde5d6b9169cfec58a8R35)? I was misled by `log logging.Logger` in function arguments since is is not underscored.

rjeczalik replied:
Done.

rjeczalik committed:
credential: remove Lock / Unlock

They should be re-introduces once more methods
are moved from provider.Builder to credential.Client.

Ref: https://github.com/koding/koding/pull/9182#discussion_r82771274

rjeczalik committed:
models: fix IsNotFound when no explicit resources are given

rjeczalik committed:
object: ensure arguments are of the same type in Merge

rjeczalik commented on go/src/koding/db/models/error.go:
```diff
@@ -0,0 +1,61 @@
+// IsNotFound gives true when err is of *NotFoundError type and its Resource
+// field is equal to one of the given resources.
+//
+// If no resources are given, the function tests only if err is of
```
Done.

rjeczalik commented on go/src/koding/kites/kloud/utils/object/merge.go:
```diff
@@ -0,0 +1,26 @@
+// Merge overwrites fields in v1 with a value of complementary
+// field in v2, if the field in v2 is non-empty.
+func Merge(v1, v2 interface{}) {
+	vv2 := reflect.ValueOf(v2).Elem()
```
Done.

rjeczalik commented on go/src/koding/kites/kloud/credential/mongo.go:
```diff
@@ -106,3 +108,470 @@ func (db *mongoStore) Put(_ string, creds map[string]interface{}) error {
+}
+
+func (db *mongoDatabase) Lock(cred *Cred) error {
+	err := db.MongoDB.Run(modelhelper.CredentialsColl, func(c *mgo.Collection) error {
```
The locking was removed for now.

rjeczalik committed:
koding/kites/config: regenerate json file

rjeczalik committed:
provider: simplify desc func

rjeczalik committed:
klientctl: log to stderr

rjeczalik committed:
credential: split Verify into smaller funcs

rjeczalik committed:
credential: reuse fetchModels in Creds method

rjeczalik commented on go/src/koding/kites/kloud/credential/mongo.go:
```diff
@@ -106,3 +108,470 @@ func (db *mongoStore) Put(_ string, creds map[string]interface{}) error {
+
+	for _, rel := range rels {
+		cred, err := modelhelper.GetCredentialByID(rel.TargetId)
+		if err == mgo.ErrNotFound {
```
Done.

rjeczalik replied:
Done.

rjeczalik commented on go/src/koding/kites/kloud/credential/mongo.go:
```diff
@@ -106,3 +108,470 @@ func (db *mongoStore) Put(_ string, creds map[string]interface{}) error {
+
+var _ Database = (*mongoDatabase)(nil)
+
+func (db *mongoDatabase) Validate(f *Filter, c *Cred) (Perm, error) {
```
Partially done, refactored, but no tests yet.

rjeczalik commented on go/src/koding/kites/kloud/credential/mongo.go:
```diff
@@ -106,3 +108,470 @@ func (db *mongoStore) Put(_ string, creds map[string]interface{}) error {
+}
+
+func (db *mongoDatabase) Creds(f *Filter) ([]*Cred, error) {
+	log := db.log().New("Creds")
```
Refactored.

rjeczalik committed:
credential: fetch all credentials at once

rjeczalik committed:
kloud: add -genschema <file> for generating provider schema file

Closes #9024.

rjeczalik commented on go/src/koding/kites/kloud/schema.go:
```diff
@@ -0,0 +1,40 @@
+package main
```
Current schema (together with Azure from https://github.com/koding/koding/pull/9241):

https://gist.github.com/rjeczalik/2785e5cab11981e174636ad2bfe6e71e

rjeczalik committed:
credential: filter out already fetched creds

rjeczalik commented:
@cihangir @ppknap Addressed most of them - missing piece are database tests for mongo, I'll send them in next PR.

ppknap commented on go/src/koding/kites/kloud/stack/provider/describe_test.go:
```diff
@@ -0,0 +1,97 @@
+	}
+
+	for name, cas := range cases {
+		desc, err := provider.Describe(cas.value)
```
just a nitpick(:smile:) - you can use `t.Run(name, func(t *testing.T){ /* ... */})` and then `t.Fatalf` instead of `t.Errorf; continue`
